src/util.ts

```diff
@@ -0,0 +1,22 @@
+export function hasProp<X extends {}, Y extends PropertyKey>(
+  obj: X,
+  prop: Y
+): obj is X & Record<Y, unknown> {
+  return prop in obj
+}
+export function hasPropType<
+  X extends {},
+  Y extends PropertyKey,
+  Z extends
+    | 'string'
+    | 'number'
+    | 'bigint'
+    | 'boolean'
+    | 'symbol'
+    | 'undefined'
+    | 'object'
+    | 'function'
+>(obj: X, prop: Y, type: Z): obj is X & Record<Y, Z> {
```

Comment:
Isn't `Z` a literal string instead of a type?

Reply:
It's a literal string representing a type.

Reply:
```ts
Record<'foo', 'string'> !== Record<'foo', string>
```

Reply:
Should I provide documentation how this util function works?

Reply:
Nope, it doesn't do what you want it to do, documenting it doesn't fix that.

Reply:
Well actually it does that very well, but perhaps it's not what you were expecting it to do? I added docs now to resolve the misunderstanding

Reply:
Okay, I think I see what you mean now. The function behaves exactly the way I want, though. I'm currently just not using it in places where the return type needs to be more accurate than `boolean`, but if someone was actually trying to utilize it in a place where flow analysis matters, it would yield the wrong type. I'll quickly fix that, thanks